### PR TITLE
#428: allow swaps of empty clusters from target

### LINKED ISSUE #428: Memory-aware LB needs to allow swaps of whole clusters with empty clusters
There may be ranks that have no clusters at all or far fewer clusters than that rank has the memory to accommodate. When attempting to swap, it should be allowed to swap a whole cluster from an overloaded rank with nothing from the underloaded rank (i.e., move a whole cluster from an overloaded to an underloaded rank even if there is nothing there to swap with).

### PR DESCRIPTION
**Pull request by ppebay**
Fixes #428

### CODE REVIEW

**Comment by ppebay** [comment]
**Key findings:**
* the initial impression that the load-balancing under-performance of the clustering-based strategy was caused by the inability to swap with empty clusters was correct;
* what was incorrect however was the first assumption that it was caused by the non-existence of a notional empty cluster on the source rank; this is because the logic being to try to offload work from overloaded to underloaded ranks, there is no point in trying to swap and empty cluster on the source in exchange for something on the trial destination; instead the symmetric exchange should be tried.

As a result, the results of the call to `__cluster_objects()` on the destination rank must always be “enriched” with the notional empty cluster, denoted as: `None: []`

**Comment by ppebay** [comment]
Example of 16 iterations of the `too-complex` case, in **non-deterministic** mode:

out_00
out_16

**N.B.:** load imbalance continues to decrease and is not blocked at that level, past iteration 16.

**Comment by ppebay** [comment]
Verifying that has no negative impact on other, previously well resolved, cases:
output_file_00
output_file_08

**Comment by ppebay** [comment]
@nlslatt should I nonetheless merge, while you are further investigating?

**Comment by nlslatt** [comment]
> @nlslatt should I nonetheless merge, while you are further investigating?

No, see Slack.

**Comment by ppebay** [comment]
Results with new execution logic (now only one target allowed to be picked but with an arbitrary numbers of swaps):

Screen Shot 2023-09-09 at 4 34 19 PM

**Comment by ppebay** [comment]
Further improved strategy, both for execution speed and load imbalance:
out_08

**Comment by ppebay** [comment]
Results also improved for the `challenging_toy_hundreds_tasks` case (also in non-deterministic mode):

output_file_00
output_file_08

File: src/lbaf/Execution/lbsClusteringTransferStrategy.py [hunk 1]
```diff
@@ -99,23 +99,33 @@ def execute(self, known_peers, phase: Phase, ave_load: float):
...
-                            break
-
-                    # Break out from targets loop once one swap was performed
-                    if swapped_cluster:
-                        break
+                        c_try = self._criterion.compute(r_src, o_src, r_try, o_try)
+                        if c_try > 0.0:
+                            # Compute source cluster size only when necessary
+                            sz_src = sum([o.get_load() for o in o_src])
+                            if  c_try > 0.05 * sz_src:
```
Comment (lifflander) [inline src/lbaf/Execution/lbsClusteringTransferStrategy.py:117]: Should this 5% be made into a tunable parameter?

Comment (ppebay) [reply src/lbaf/Execution/lbsClusteringTransferStrategy.py:117]: Yes. I am creating a follow-on issue for this.

Comment (ppebay) [reply src/lbaf/Execution/lbsClusteringTransferStrategy.py:117]: @lifflander @cwschilly #434

**Comment by nlslatt** [comment]
@PPP This is working better for me as well

**Comment by ppebay** [comment]
Thanks @nlslatt @lifflander , can we merge this PR?

File: src/lbaf/Execution/lbsClusteringTransferStrategy.py [hunk 2]
```diff
@@ -99,23 +99,33 @@ def execute(self, known_peers, phase: Phase, ave_load: float):
...
+                                self._logger.info(
+                                    f"Swapping cluster {k_src} of size {sz_src} with cluster {k_try} on {r_try.get_id()}")
+                                self._n_transfers += phase.transfer_objects(
+                                    r_src, o_src, r_try, o_try)
+                                n_swaps += 1
+                                break
+
+                # In non-deterministic mode break out from targets as soon as swaps occurred
+                if n_swaps and not self._deterministic_transfer:
+                    break
```
Comment (nlslatt) [inline src/lbaf/Execution/lbsClusteringTransferStrategy.py:128]: Can you please explain the purpose of the two different `break` statements in this outer for loop?

**Review by nlslatt — COMMENTED** [review]
There is also a commit message that is missing the issue number.

File: src/lbaf/Execution/lbsClusteringTransferStrategy.py (hunk 2 above)
Comment (ppebay) [reply src/lbaf/Execution/lbsClusteringTransferStrategy.py]: Thanks for the good catch @nlslatt , lines 126--128 are indeed a leftover debug code that I have now deleted. Thanks!

**Review by nlslatt — APPROVED** [review]
Looks good

**Pull request merged**